File: IF_3.1/LK_Info_Flow/examples/case1_bivariate_causality.py
# case 1 (bivariate causality):
import numpy as np
np.random.seed(0)  
from LK_Info_Flow import causal
#import matplotlib.pyplot as plt

#generate dataset
a11=0.3;a12=-0.4;a22=0.7;a21=0;b1=0.5;b2=0.5;
x=np.zeros([10000,])
y=np.zeros([10000,])
x[0]=0.4;y[0]=0.3;
for i in range(10000-1):
    x[i+1]=a11*x[i]+a21*y[i]+b1*np.random.normal();
    y[i+1]=a12*x[i]+a22*y[i]+b2*np.random.normal();

#print the structure of system
print ("x(i+1)=%.2f * x(i) + %.2f * y(i) + %.2f W" % (a11, a21,b1))
print ("y(i+1)=%.2f * x(i) + %.2f * y(i) + %.2f W" % (a12, a22,b2))


#initialization
X=np.ones((2,10000));
X[0]=x;X[1]=y;

#calculate the causality
causal_graph=causal.multi_causality_est_OLS(X[:,200:]);


#information flow
IF=np.squeeze(causal_graph.get('IF'));

#normalized information flow
nIF=np.squeeze(causal_graph.get('nIF'));

#significant test 
e99=np.squeeze(causal_graph.get('err_e99'));

## plot heatmap
# print(nIF[0,1])
# plt.matshow(nIF, 0,cmap='RdYlGn',vmin=-0.01,vmax=0.01)


# show the results
if abs(IF[0,1])>e99[0,1]:
    print('x -> y percent:%5.2f' % (nIF[0,1]*100)+'%')
else:
    print('x not -> y')


if abs(IF[1,0])>e99[1,0]:
    print('y -> x precent:%5.2f' % (nIF[1,0]*100)+'%')
else:
    print('y not -> x')

File: codes/R/case1_bivariate_causality.R
# case 1 (bivariate causality):
set.seed(0)
source("LK_Info_Flow.R") 

#generate dataset

a11 <- 0.3
a12 <- -0.4
a22 <- 0.7
a21 <- 0
b1 <- 0.5
b2 <- 0.5

x <- rep(0, 10000)
y <- rep(0, 10000)
x[1] <- 0.4
y[1] <- 0.3

#simulate the time series
for(i in 1:(length(x)-1)){
 x[i+1] <- a11 * x[i] + a21 * y[i] + b1 * rnorm(1)
 y[i+1] <- a12 * x[i] + a22 * y[i] + b2 * rnorm(1)
}

#print the structure of system
cat(sprintf("x(i+1)=%.2f * x(i) + %.2f * y(i) + %.2f W\n", a11,a21,b1))
cat(sprintf("y(i+1)=%.2f * x(i) + %.2f * y(i) + %.2f W\n", a12,a22,b2))

#initialization
#X <- matrix(c(x, y), nrow = 2)  
#
X <- rbind(x,y)

## write.table(X, file = "case1_data.txt", sep = "\t", col.names = F, row.names = F)
X <- as.matrix(read.table("case1_data.txt", header = FALSE))

X1=X[1:2,201:10000] 
#calculate the causality
causal_graph <- multi_causality_est(X1)

#information flow
IF <- causal_graph$IF[,,1]

#normalized information flow
nIF <- causal_graph$nIF[,,1]

#significant test
e99 <- causal_graph$err_e99[,,1]

if (abs(IF[1,2]) > e99[1,2]) {
  cat(paste('y -> x percent:', format(nIF[1,2]*100, nsmall = 2), '%\n'))
} else {
  cat('y not -> x\n')
}

if (abs(IF[2,1]) > e99[2,1]) {
  cat(paste('x -> y percent:', format(nIF[2,1]*100, nsmall = 2), '%\n'))
} else {
  cat('x not -> y\n')
}